### ISSUE TITLE
cmd/tailscaled: tracking bug for ongoing log leveling work

### ISSUE THREAD

**Issue by bradfitz** [tailscale/tailscale#1548]
We introduced log levels and fixed the bulk of the log spam issues in #282 but work remains.

This is an ongoing tracking bug for future log leveling work.

And for tailscaled flags to control log output/uploading.

**Comment by luismanson** [comment]
Hello, is this active? I have been poking with syslog on a Raspberry Pi and noticed Tailscale logging being too verbose.
Is some form of logging control comming?

**Comment by DentonGentry** [comment]
Substantial work on reducing log verbosity has already been done, though there is more that can be done and would be tracked here. There is not currently a large effort to further reduce overall logging, though there are a couple places where the code gets into a state where it logs stuff too often. Those are being resolved by fixing the state it can get into.

**Comment by josharian** [comment]
@luismanson are there particular log entries that are of concern, or is this a general "there's too much"?

**Comment by luismanson** [comment]
Thanks for your quick reply @DentonGentry 

@josharian  thanks for your interest,  in my case I've seen my 7Mb 'daemon.log" is like >30% filled with tailscale related messaged, most of them are some variation of:
`"tailscaled[423]: Accept: TCP{123.123.193.26:45542 > 123.123.159.56:22} 104 tcp non-syn`

Some quick numbers:
$ grep -i tailscale daemon.log|grep -v Accept|wc -l
    4340
$ grep -i tailscale daemon.log|grep Accept|wc -l
    16118

**Comment by dijit** [comment]
Tailscale logging is indeed rather verbose; in fact it's the kind of output I would expect if I ran it with the `--verbose` flag, since every connection, drop, health status and so on is a log line.

sample from my laptop (note: I didn't have to remove any lines from any _other_ service on my laptop)

```May 21 16:40:39 trace.rundeln.drk.sc tailscaled[438]: health("overall"): error: not in map poll
May 21 16:40:40 trace.rundeln.drk.sc tailscaled[438]: [RATELIMIT] format("control: mapRoutine: %s") (2 dropped)
May 21 16:40:40 trace.rundeln.drk.sc tailscaled[438]: control: mapRoutine: state:authenticated
May 21 16:40:40 trace.rundeln.drk.sc tailscaled[438]: trying bootstrapDNS("derp7.tailscale.com", "167.179.89.145") for "login.tailscale.com" ...
May 21 16:40:41 trace.rundeln.drk.sc tailscaled[438]: bootstrapDNS("derp7.tailscale.com", "167.179.89.145") for "login.tailscale.com" = [2a05:d014:386:203:f8b4:1d5a:f163:e187 3.121.18.47]
May 21 16:40:41 trace.rundeln.drk.sc tailscaled[438]: [RATELIMIT] format("Received error: %v") (3 dropped)
May 21 16:40:41 trace.rundeln.drk.sc tailscaled[438]: Received error: PollNetMap: Post "https://login.tailscale.com/machine/3ca893fc9fb626d574bd53b74226a0a119224661fc0eb4756bbd9376b150c61f/map": dial tcp [2a05:d014:386:203:f8b4:1d5a:f163:e187]:443: connect: network is unreachable
May 21 16:40:41 trace.rundeln.drk.sc tailscaled[438]: [RATELIMIT] format("control: %s: backoff: %d msec") (3 dropped)
May 21 16:40:41 trace.rundeln.drk.sc tailscaled[438]: control: mapRoutine: backoff: 2983 msec
May 21 16:40:44 trace.rundeln.drk.sc tailscaled[438]: control: mapRoutine: state:authenticated
May 21 16:40:44 trace.rundeln.drk.sc tailscaled[438]: [RATELIMIT] format("control: mapRoutine: %s")
May 21 16:40:44 trace.rundeln.drk.sc tailscaled[438]: trying bootstrapDNS("derp1.tailscale.com", "159.89.225.99") for "login.tailscale.com" ...
May 21 16:40:44 trace.rundeln.drk.sc tailscaled[438]: bootstrapDNS("derp1.tailscale.com", "159.89.225.99") for "login.tailscale.com" = [2a05:d014:386:203:f8b4:1d5a:f163:e187 3.121.18.47]
May 21 16:40:44 trace.rundeln.drk.sc tailscaled[438]: Received error: PollNetMap: Post "https://login.tailscale.com/machine/3ca893fc9fb626d574bd53b74226a0a119224661fc0eb4756bbd9376b150c61f/map": dial tcp [2a05:d014:386:203:f8b4:1d5a:f163:e187]:443: connect: network is unreachable
May 21 16:40:44 trace.rundeln.drk.sc tailscaled[438]: [RATELIMIT] format("Received error: %v")
May 21 16:40:44 trace.rundeln.drk.sc tailscaled[438]: control: mapRoutine: backoff: 2181 msec
May 21 16:40:44 trace.rundeln.drk.sc tailscaled[438]: [RATELIMIT] format("control: %s: backoff: %d msec")
May 21 16:40:46 trace.rundeln.drk.sc tailscaled[438]: trying bootstrapDNS("derp4.tailscale.com", "167.172.182.26") for "login.tailscale.com" ...
May 21 16:40:46 trace.rundeln.drk.sc tailscaled[438]: bootstrapDNS("derp4.tailscale.com", "167.172.182.26") for "login.tailscale.com" = [2a05:d014:386:203:f8b4:1d5a:f163:e187 3.121.18.47]
May 21 16:40:51 trace.rundeln.drk.sc tailscaled[438]: [RATELIMIT] format("control: mapRoutine: %s") (1 dropped)
May 21 16:40:51 trace.rundeln.drk.sc tailscaled[438]: control: mapRoutine: state:authenticated
May 21 16:41:00 trace.rundeln.drk.sc tailscaled[438]: health("overall"): ok
May 21 16:43:23 trace.rundeln.drk.sc tailscaled[438]: logtail: dialed "log.tailscale.io:443" in 170ms
May 21 16:43:27 trace.rundeln.drk.sc tailscaled[438]: netmap diff:
May 21 16:43:27 trace.rundeln.drk.sc tailscaled[438]: - [MNw8B] d:2e4d5ea364975ef4 D4 100.101.102.103 fd7a:115c:a1e0:ab12:4843:cd96:6265:6667/128 :   18.196.71.179:41641      172.17.0.1:41641   172.31.34.125:41641
May 21 16:43:27 trace.rundeln.drk.sc tailscaled[438]: + [MNw8B] d:2e4d5ea364975ef4 D4 100.101.102.103 fd7a:115c:a1e0:ab12:4843:cd96:6265:6667/128 :      172.17.0.1:41641   172.31.34.125:41641```

lines such as `+ [MNw8B] d:2e4d5ea364975ef4 D4 100.101.102.103 fd7a:115c:a1e0:ab12:4843:cd96:6265:6667/128 :      172.17.0.1:41641   172.31.34.125:41641` are effectively meaningless to me and I presume they help debugging if there's an issue.

`logtail: dialed "log.tailscale.io:443" in 170ms` is informational

`health("overall"): ok` is information/verbose

`bootstrapDNS("derp4.tailscale.com", "167.172.182.26") for ` also informational

`Received error: PollNetMap: Post "https://login.tailscale.com/machine/3ca893fc9fb626d574bd53b74226a0a119224661fc0eb4756bbd9376b150c61f/map": dial tcp [2a05:d014:386:203:f8b4:1d5a:f163:e187]:443: connect: network is unreachable` is the kind of thing I would expect to see, in my logs, it indicates "something unexpected happened", The default log level for people should always be that we only log on unexpected outcomes or things which may severely alter the running state.

Often it's nice to hide this level of complexity away in a `--verbose` flag.

**Comment by luismanson** [comment]
Any updates on this? 
Logging still is very verbose, we need some flag to control or disable it, current level looks like a DEBUG.

Thanks

**Comment by kgersen** [comment]
same here i'm getting ~2000 entries per hour in the journal:

```
sudo journalctl --since "1h ago" -t tailscaled | wc -l
1930
```

```
sudo journalctl --since "24h ago" -t tailscaled | wc -l
46424
```

```
sudo journalctl --since "1h ago" -t tailscaled | cut -f6 -d" " | sort | uniq -c
      4 +
      4 -
      1 2019-12-26
    360 Accept:
      8 derphttp.Client.Recv:
      9 derphttp.Client.Send:
    111 magicsock:
      4 netmap
    721 open-conn-track:
    716 [RATELIMIT]
```

so mostly `RATELIMIT` and `open-conn-track` events.

**Comment by josharian** [comment]
@kgersen a bit more detail would be useful to see why those particular lines are occurring so much for you. Would you run [`tailscale bugreport`](https://tailscale.com/kb/1080/cli/?q=Cli#bugreport) and give me the bug report identifier? Thanks.

**Comment by kgersen** [comment]
> @kgersen a bit more detail would be useful to see why those particular lines are occurring so much for you. Would you run [`tailscale bugreport`](https://tailscale.com/kb/1080/cli/?q=Cli#bugreport) and give me the bug report identifier? Thanks.

BUG-f3b5af8c15d455139c492d0bbd92d9cf59a6db6718fde0977599b351e3fd9257-20210918100211Z-19f4c83f8ab733f0

**Comment by bradfitz** [comment]
@kgersen, that Arch machine is running some custom build that doesn't even advertise its version so it's a little hard to debug well.

But most of the noise is due to it constantly trying to reach these two peers and getting TCP RSTs from them:

```
2021-09-17 00:42:39.306436376 +0200 +0200: open-conn-track: flow TCP (TCP 100.96.126.125:39040 => 100.108.221.119:9100) got RST by peer
2021-09-17 00:42:41.203488606 +0200 +0200: open-conn-track: flow TCP (TCP 100.96.126.125:40812 => 100.65.173.41:9323) got RST by peer
```

In any case, a separate bug is probably warranted to discuss that further so we don't distract from this meta issue.

**Comment by kgersen** [comment]
@bradfitz  tailscale is installed with 'pacman' nothing custom. I've updated it just now.
```
tailscale version
1.14.3
  tailscale commit: a5b1456410a24519fc11cb49d4f43b7b7befbee4
  go version: go1.17.1

```
so it's https://archlinux.org/packages/community/x86_64/tailscale/

These reset are normal. it's just Prometheus that is trying to scrape endpoints that don't exist all the time (no process is listening at that port).

I don't think this should be logged at the tailscale level or there should be a way to not log failed TCP connections.

step to reproduce:
`watch -n 1 curl http://another_tailscale_node:12345`
with no process listening at port 12345, this will fill the journal with these messages (open-conn-track then RATELIMIT).

**Comment by wogong** [comment]
1+ years passed, you still can't fix this? plz don't ruin my journal.

**Comment by kgersen** [comment]
quick fix: remove/comment the 3 lines at https://github.com/tailscale/tailscale/blob/7bfd4f521d9078caba692f33a04b288f1c453ab7/wgengine/pendopen.go#L97

**Comment by kverb** [comment]
it's a shame we can't use an env var or command line flag to decrease logging verbosity. In case it helps any loki + promtail users, you can filter out tailscale logs from the journald scraper with the 'drop' action in promtail.yaml's `relabel_configs:`:

```   relabel_configs:
      - source_labels: ['__journal__systemd_unit']
        target_label: 'unit'
      - action: drop
        regex: '.*tailscale.*'
        source_labels: ['__journal__systemd_unit', 'unit']```

The actual journal is still poluted, but at least they aren't spewing into loki now.

**Comment by 33b5e5** [comment]
I find the excessive logging annoying also. It's disproportionate in volume compared to all the other services running on my Linux boxes. I'm also slightly uncomfortable with the notion of all this debug-level data being shipped off to an AWS node. More control over this would be good.

In the mean time, I've discovered I can squelch all the logging on Ubuntu/Debian by editing the tailscaled systemd unit with an override that redirects the logs to /dev/null:

`sudo systemctl edit tailscaled.service`

```
### Editing /etc/systemd/system/tailscaled.service.d/override.conf
### Anything between here and the comment below will become the new contents of the file

[Service]
StandardOutput=null
```

**Comment by luismanson** [comment]
Hello, may I ask if this is being discussed/fixed anytime soon?

**Comment by xiaolei0125** [comment]
Hello, any update about this issue?  
too much tailscaled log in my SD card.

**Comment by luismanson** [comment]
Thanks @jbosboom what version `journalctl --version` do you have? the namespace flag is not on my Debian w/ systemd 241
Anyways it's been more than a year and it's clear logging control is not a priority, this makes Tailscale a very bad option for a raspberry w/ SD card and maybe some other embedded systems,

**Comment by DentonGentry** [comment]
Related: Tailscale 1.32 added a `--no-logs-no-support` argument to `tailscaled`

**Comment by DentonGentry** [comment]
When we have further updates we will post them to the bug, there is nothing further to report right now. You're welcome to redirect stdout to /dev/null if that solves a pain point.

**Comment by sedlund** [comment]
for nix

```nix
 systemd.services.tailscaled.serviceConfig = {StandardOutput = "null";};
```

**Comment by tshcherban** [comment]
Any updates with logging?

**Comment by SamyDjemai** [comment]
Hi, better log leveling would be great to have. We're running lots of Tailscale nodes as Kubernetes pods, and we've had to setup the following Datadog exclusion filter:
```
container_name:(tailscale OR ts-sidecar) ("wgengine: Reconfig:" OR "dns: Resolvercfg:" OR "dns: OScfg: {}" OR "dns: Set:" OR "[RATELIMIT]") 
```

There are nearly 15 million logs matching this filter, from 40 Tailscale containers. An env variable that would preemptively suppress these logs would be really nice to have.

**Comment by davidemyers** [comment]
FWIW, I have the following in a script I use to configure Tailscale on Ubuntu Server. This way I keep the logs around in plain files "the old-fashioned way" rather than in the systemd journal.

```bash
cat > /etc/systemd/system/tailscaled.service.d/override.conf << EOF
[Service]
StandardOutput=append:/var/log/tailscaled.log
StandardError=append:/var/log/tailscaled.log
EOF

cat > /etc/logrotate.d/tailscale << EOF
/var/log/tailscaled.log
{
    rotate 4
    copytruncate
    weekly
    missingok
    notifempty
    compress
    delaycompress
}
EOF

systemctl daemon-reload
systemctl restart tailscaled.service
```

**Comment by SuperSandro2000** [comment]
IMO systemd-journal is the best things since sliced bread. No longer can a server's rootfs run full because of an application that just logs way to much as journald will just rotate it away automatically.

**Comment by anacrolix** [comment]
Adding a use case that's not well represented in the existing discussion: **Docker containers**.

In Docker, log output goes through the container log driver (json-file, journald, fluentd, etc.) with no rate limiting by default. Tailscale's log volume makes it the dominant source of log data by a large margin, causing real problems for anyone shipping container logs to a centralized store — the cost and storage impact is significant across many containers.

The **userspace networking mode** () compounds this: it's the required mode for running Tailscale as a sidecar or in unprivileged containers, but it generates additional log noise, and the documentation doesn't explain the tradeoffs or how to reduce verbosity in that mode.

Some kind of  or  flag (separate from the logtail upload verbosity) would go a long way here. Even just suppressing debug-level output to stderr/stdout by default would help Docker users significantly.

**Comment by CharlesScoville** [comment]
FYI: systemd-journald has a (largely unacknowledged) write amplification "issue" (bug) since about 2021 that puts some 10x more writes down to the disk than one might expect, making this problem more than 10x larger than one would expect.

TL;DR: Being able to turn this way down or off isn't just a convince, it saves flash memory devices from a premature death.
